pkg/Library/NodeGraph/NodeTree.js

```diff
@@ -133,12 +140,38 @@ onSelect({eventlet: {value: container}}, state) {
   state.selectedContainer = container;
 },
 
-async onSetContainer({selectedNode: node}, {selectedContainer: container}, {service}) {
+async onSetContainer({selectedNode, pipeline}, {selectedContainer: container}, {service}) {
   if (container) {
-    await service({kind: 'ComposerService', msg: 'setContainer', data: {node, container}});
+    const hosts = selectedNode?.position?.preview;
+    const hostId = hosts ? Object.keys(hosts).pop().split(':')?.[0] : '';
+    await service({kind: 'ComposerService', msg: 'setContainer', data: {hostId, container}});
+    selectedNode = this.updateContainerInNode(selectedNode, hostId, container);
+    return {
+      selectedNode,
+      pipeline: this.updateNodeInPipeline(selectedNode, pipeline)
+    };
   }
 },
 
+updateContainerInNode(node, hostId, container) {
+  return {
```

Comment:
same as line 163, I'm not sure you need to construct a whole new node here

Reply:
wdym? i tried instead:
```
node.position = node.position || {};
node.position.preview = node.position.preview || {};
node.position.preview[`${hostId}:Container`] = container;
return {...node};
```

but it didn't work.

Reply:
"it didn't work" is not enough info :shrug:
mysteries are not allowed 😃 

`return {...node}` is "making a new instance", and I don't know why that's better than `return node`. This is the point, use whatever syntax you want, just avoid creating copies of things for no reason.

Reply:
Side note: let's be careful about magic strings like `${hostId}:Container` ... in this case, probably best to put it out-of-line (i.e. `const key = \`${hostId}:Container\``)

Reply:
from what I observe, the pipelines aren't persisted properly without the copying.
moreover, the change i made in `updateNodeInPipeline` (where i set `pipeline.nodes[index] = node;` instead of overriding the array) is affecting that as 

the screenshot is showing that the persisted data and the runtime data in devtools are different:
https://screenshot.googleplex.com/A49FYNcWYVepVk6.png

as a result, the `container` info is lost on refresh. i'll keep debugging, but these are my observations so far.